Fix TwistedSystem (plugin for pydoctor)

mthuurne opened the pull request:
## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10044
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I ran additional checks listed at [Getting Your Patch Accepted](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have updated the automated tests.
* [ ] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

mthuurne committed:
Fix computation of full name of decorator in pydoctor plugin

The full name was computed incorrectly if it contained dots, meaning
`@deprecated` worked as expected while `@deprecate.deprecated` did not.

Using `node2fullname()` is also safer, since it deals with situations
where `node2dottedname()` returns `None`.

mthuurne committed:
Fix TwistedModuleVisitor

When pydoctor switched from using the `compiler` module to `ast`,
the names of the visitor methods changed.

Recent pydoctor versions ignore inner functions, so we can't assume
that the current node will always be added to `contents`.

Also, don't bother making a copy of the decorators, since the
getDeprecated() function does not modify the list.

mthuurne committed:
Remove TwistedFunction

This was an unnecessary hack to generate calls to getDeprecated().

mthuurne committed:
Fix skipping of release tests

These were always skipped because every value other than None for
the 'skip' attribute was evaluated as True, even the empty string.

Based on work @adiroiban did in #1466.

mthuurne committed:
Remove default encoding ("utf-8")

mthuurne committed:
Encode/decode documentation output strings

These tests weren't ported to Python 3 yet, but no-one noticed because
of the skip bug.

mthuurne committed:
Do not re-raise AttributeError in getDeprecated()

I think this was debug code that accidentally got committed: if it was
deliberate, the whole `try:` block could have been removed instead.

It is possible for AttributeError to be raised when something other
than a string literal node is passed to versionToUsefulObject().

mthuurne committed:
Compare captured stdout as bytes

mthuurne committed:
Support 'replacement' as a keyword argument of '@deprecated'

This is used in both the unit test and in Twisted's code, but wasn't
supported yet by the pydoctor plugin.

mthuurne committed:
Add empty newsfragment

adiroiban commented on src/twisted/python/test/test_release.py:
```diff
@@ -912,15 +877,26 @@ def test_export(self):
+    if int(gitVersion[0]) >= 2:
+        gitSkip, gitSkipText = False, ""
+    else:
+        gitSkip, gitSkipText = True, "old git is present"
```
I think that in these days all the supported environment if they have git, will have a git older than v2


And maybe we can also assume that all testing system have git... so we can always run it.

I think that this might be a reason why the tests were skipped on Windows, as getting the git path on Windows might be a little bit more complicated.

mthuurne replied:
One reason it failed on Windows was that `/tmp` was hardcoded as the temp dir. That's easy to fix though.

Windows using "git-bash" (msys) is actually close enough to a POSIX environment for a lot of tasks. But I don't know if our CI runs under bash or `cmd.exe`.

mthuurne replied:
After fixing the temp dir, more issues are popping up. So I disabled the git tests on Windows.

Once the pydoctor and sphinx tests are running in CI, we can see whether those do run correctly on Windows or should be disabled as well.

mthuurne committed:
Use tempfile.gettempdir() instead of hardcoding /tmp

mthuurne committed:
Remove 'docstring' argument to Module/Package constructors

This argument does not exist anymore. Given that the value 'ignored'
was passed by the tests, removing the arguments shouldn't change
their effectiveness.

mthuurne committed:
Disable git release tests on Windows

There are issues with building the command line and removal of
temporary directories. Since these tests were excluded before,
it should be fine to exclude them again.

adiroiban commented on setup.cfg:
```diff
@@ -41,6 +41,8 @@ packages = find:
 test =
     cython-test-exception-raiser ~= 1.0
     PyHamcrest >= 1.9.0
+    pydoctor >= 20.7.0
```
Thanks for the great start.


The updates to `test` are not quite ok.

In Twisted we are running tests without dependencies to make sure that twisted can handle soft dependencies.

We should not add more deps to this... and the long term plan is to also remove pyhamcrest or skip hamcrest test when pyhamcrest is not installed.

Yes... this did the trick, but I think that we should add a dev_docs extra requirements.

I will try to see if I can make it right  :)

adiroiban committed:
Create dedicated extra deps for documentation and release.

adiroiban committed:
Move mock package back to dev.

adiroiban commented:
Test are failing on windows due to the source path URL.
On windows we end up with: https://github.com/twisted/twisted/tree/twisted-1.0.0/src\\twisted

I think that the issue is here:

```
            mod.sourceHref = self.sourcebase + source_path[len(projBaseDir):]
```

https://github.com/twisted/pydoctor/blob/678441a505b4ee8700b29dfd6c264e1d21e294a8/pydoctor/model.py#L572

we should do a string replace on source_path and replace \ with / to normalize the URL

adiroiban reviewed (approved):
This looks good.

I can see that test_buildWithDeprecated is now executed and is green

https://github.com/twisted/twisted/pull/1473/checks?check_run_id=1323782792#step:8:7444

Some tests on windows are failing ... and I think that due to pydoctor.

Can we fix pydoctor and do a new release of pydoctor ?

adiroiban committed:
Skip all release test on Windows.

adiroiban commented:
So... my plan for this PR is the following:

* Merge with skiping all release tests on windows.
* Fix pydoctor source URL on windows and release a new version of pydoctor
* Update trial to skip module on empty string so that we don't need the typing comment
* Have a separate PR in which we use the updated pydoctor and enable the release tests on Windows.

 I see that release tests are executed on Linux https://github.com/twisted/twisted/pull/1473/checks?check_run_id=1323917764#step:8:7473

@mthuurne If you are happy with the plan feel free to merge this.

Thanks for your help! :)

graingert commented on setup.cfg:
```diff
@@ -38,25 +38,37 @@ package_dir = =src
+    twisted-dev-tools >= 0.0.2
+    python-subunit
+    twistedchecker >= 0.7.2
+    mock~=4.0
```
mock is for readthedocs

adiroiban replied:
Thanks.  I will move it...

But do we need to pin the mock value?

I guess that we only need to pin the high level read the docs and that package will pull its dependencies. Isn't that the case?

Thanks!

adiroiban replied:
Thanks. I tried to simplify the dependencies. Since these are dev and release only tools I think that is ok to be more flexible with the dependencies and don't pin everything.


I see that Read The docs build is successful with relaxed dependencies :)

adiroiban committed:
Update dev release dependencies.

graingert commented on setup.cfg:
```diff
@@ -38,25 +38,30 @@ package_dir = =src
-    towncrier >= 17.4.0
     twistedchecker >= 0.7.2
-    # force upgrades for rtd default packages: https://git.io/JU73V
-    alabaster~=0.7.12
-    commonmark~=0.9.1
-    docutils~=0.16.0
-    mock~=4.0
-    pillow~=7.2
-    readthedocs-sphinx-ext~=2.1
-    recommonmark~=0.6.0
-    sphinx~=3.2
-    sphinx-rtd-theme~=0.5.0
```
so we're going to be using the old deps now? https://github.com/readthedocs/readthedocs.org/blob/046c4bd20fd875807cb3e1840846249160c37079/readthedocs/doc_builder/python_environments.py#L354-L389

graingert replied:
```
readthedocs=
    %(dev_release)s
    # force upgrades for rtd default packages: https://git.io/JU73V
    alabaster~=0.7.12
    commonmark~=0.9.1
    docutils~=0.16.0
    mock~=4.0
    pillow~=7.2
    recommonmark~=0.6.0
```

and then update the `extra` in `./readthedocs.yml`

graingert commented on src/twisted/python/test/test_release.py:
```diff
@@ -47,42 +50,9 @@
 )
 
 if os.name != "posix":
```
you should be able to get mypy to infer the type with:
```python
skip = (
    None
    if os.name == "posix"
    else "Release toolchain only supported on POSIX."
)
```

also I think `sys.platform != "win32"` is preferred?

https://gitter.im/python-trio/general?at=5da80653c87a1d28acc46108

adiroiban replied:
Thanks. I will apply the changes

graingert commented on src/twisted/python/test/test_release.py:
```diff
@@ -912,15 +883,29 @@ def test_export(self):
         self.assertStructure(exportDir, structure)
 
 
+if os.name != "posix":
+    gitSkip, gitSkipText = True, "Release toolchain only supported on POSIX."
+else:
+    try:
+        gitVersion = runCommand(["git", "--version"]).split(b" ")[2].split(b".")
+    except FileNotFoundError:
+        gitSkip, gitSkipText = True, "git is not present."
+    else:
+        # We want git 2.0 or above.
+        if int(gitVersion[0]) >= 2:
+            gitSkip, gitSkipText = False, ""
+        else:
+            gitSkip, gitSkipText = True, "old git is present"
```
I think this could be more clear

```python
def _gitSkip() -> str:
    if os.name != "posix":
        return "Release toolchain only supported on POSIX."
    try:
        gitVersion = runCommand(["git", "--version"]).split(b" ")[2].split(b".")
    except FileNotFoundError:
        return "git is not present"
    if int(gitVersion[0]) < 2:
        return "git 2.0 or above is required"
    return ""

gitSkipText = _gitSkip()
gitSkip = bool(gitSkipText)
```

adiroiban replied:
I  went with removing the whole git existence check.

I think that any env in which the tests are executed will have git...and will have a newer version of git.

adiroiban committed:
Cleanup after review.

adiroiban committed:
Reformat file.

adiroiban committed:
Fix skip condition.

adiroiban committed:
Remove unused import.
